src/main/java/io/supertokens/session/accessToken/AccessTokenSigningKey.java

```diff
@@ -69,142 +79,227 @@ public static AccessTokenSigningKey getInstance(Main main) {
+
+            // start transaction
+            sqlStorage.startTransaction(con -> {
+                KeyValueInfo legacyKey = sqlStorage.getLegacyAccessTokenSigningKey_Transaction(con);
+
+                if (legacyKey != null) {
+                    sqlStorage.addAccessTokenSigningKey_Transaction(con, legacyKey);
+                    sqlStorage.removeLegacyAccessTokenSigningKey_Transaction(con);
+                    sqlStorage.commitTransaction(con);
+                }
+                return legacyKey;
+            });
+        } else {
+            SessionNoSQLStorage_1 noSQLStorage = (SessionNoSQLStorage_1) storage;
+            KeyValueInfoWithLastUpdated legacyKey = noSQLStorage.getLegacyAccessTokenSigningKey_Transaction();
+
+            if (legacyKey != null) {
+                // We should only get here once, after an upgrade.
+                // We avoid adding duplicates by enforcing that a legacy key can only ever be
+                // the first one in the new table.
+                noSQLStorage.addAccessTokenSigningKey_Transaction(
+                    new KeyValueInfo(legacyKey.value, legacyKey.createdAtTime), null
+                );
+                // We don't need to check the lastUpdatedSign here, since we never update or set
+                // legacy keys anymore.
+                noSQLStorage.removeLegacyAccessTokenSigningKey_Transaction();
+            }
+        }
+    }
+
+    public synchronized void cleanExpiredAccessTokenSigningKeys() throws StorageQueryException {
+        SessionStorage storage = StorageLayer.getSessionStorage(main);
+        CoreConfig config = Config.getConfig(main);
+
+        if (config.getAccessTokenSigningKeyDynamic()) {
+            final long signingKeyLifetime = config.getAccessTokenSigningKeyUpdateInterval()
+                    + SIGNING_KEY_VALIDITY_OVERLAP * config.getAccessTokenValidity();
+
+            storage.removeAccessTokenSigningKeysBefore(System.currentTimeMillis() - signingKeyLifetime);
+        }
+    }
+
+    public synchronized List<KeyInfo> getAllKeys() throws StorageQueryException, StorageTransactionLogicException {
+        if (this.validKeys == null || this.validKeys.size() == 0) {
+            this.validKeys = maybeGenerateNewKeyAndUpdateInDb();
         }
 
         CoreConfig config = Config.getConfig(main);
-        if (config.getAccessTokenSigningKeyDynamic() && System.currentTimeMillis() > (this.keyInfo.createdAtTime
-                + config.getAccessTokenSigningKeyUpdateInterval())) {
+
+        // If a key in the cache has expired we should refresh the list.
+        if (this.validKeys.stream().anyMatch(((KeyInfo k) -> k.expiryTime < System.currentTimeMillis()))) {
+            this.validKeys = maybeGenerateNewKeyAndUpdateInDb();
+        }
+
+        if (config.getAccessTokenSigningKeyDynamic()
+                && System.currentTimeMillis() > (this.validKeys.get(0).createdAtTime
+                        + config.getAccessTokenSigningKeyUpdateInterval())) {
```

Comment:
`maybeGenerateNewKeyAndUpdateInDb` should be called exactly once here, so all 3 if statements could be combined into a single one.
We can filter keys first, but `this.validKeys.size() == 0` should never be true otherwise. If it's simpler that way I'll update it.